pkg/cloudprovider/providers/oci/fault_domain_controller.go

```diff
@@ -0,0 +1,149 @@
+package oci
+
+import (
+	"context"
+	"fmt"
+	"go.uber.org/zap"
+	"k8s.io/client-go/util/workqueue"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/oracle/oci-go-sdk/core"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+)
+
+var UpdateNodeSpecBackoff = wait.Backoff{
+	Steps:    20,
+	Duration: 50 * time.Millisecond,
+	Jitter:   1.0,
+}
+
+type FaultDomainController struct {
+	nodeInformer coreinformers.NodeInformer
+	kubeClient clientset.Interface
+	recorder record.EventRecorder
+	cloud *CloudProvider
+	queue workqueue.RateLimitingInterface
+	logger *zap.Logger
+}
+
+// NewFaultDomainController creates a FaultDomainController object
+func NewFaultDomainController(
+	nodeInformer coreinformers.NodeInformer,
+	kubeClient clientset.Interface,
+	cloud *CloudProvider) *FaultDomainController {
+
+	logger := zap.L()
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "fault-domain-controller"})
+	eventBroadcaster.StartLogging(glog.Infof)
+	if kubeClient != nil {
+		logger.Info("Sending events to api server.")
+		eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	} else {
+		logger.Info("No api server defined - no events will be sent to API server.")
+	}
+
+	fdc := &FaultDomainController{
+		nodeInformer: nodeInformer,
+		kubeClient:   kubeClient,
+		recorder:     recorder,
+		cloud:        cloud,
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		logger: 	  logger,
+	}
+
+	// Use shared informer to listen to add nodes
+	fdc.nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			node := obj.(*v1.Node)
+			fdc.queue.Add(node.Name)
+		},
+	})
+
+	return fdc
+}
+
+func (fdc *FaultDomainController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	defer fdc.queue.ShutDown()
+
+	fdc.logger.Info("Starting fault domain controller")
+
+	if !cache.WaitForCacheSync(stopCh, fdc.nodeInformer.Informer().HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
+	}
+
+	wait.Until(fdc.runWorker, time.Second, stopCh)
```

Comment:
nit: since we are utilising workqueue, we can expedite consuming objects via more worker threads.